connector/src/main/scala/com/microsoft/kusto/spark/datasource/KustoRelation.scala

```diff
@@ -45,7 +50,15 @@ private[kusto] case class KustoRelation(kustoCoordinates: KustoCoordinates,
 
   override def buildScan(): RDD[Row] = {
     val kustoClient = KustoClientCache.getClient(kustoCoordinates.cluster, authentication).engineClient
-    val count = KDSU.countRows(kustoClient, query, kustoCoordinates.database)
+
+    var count = 0
+    try {
+      count = Await.result(Future(KDSU.countRows(kustoClient, query, kustoCoordinates.database)), timeout)
+    } catch {
+      // Count must be high if took more than 10 seconds
+      case TimeoutException => count = KustoConstants.directQueryUpperBoundRows + 1
```

Comment:
this is really confusing. If what you want is to prevent lean mode from running, please use a boolean inseam, and should not be aware of internal logic of `isLeanReadMode` (essentially you are hacking code you shouldn't be aware of here).

I would lift the `useLeanMode` flag above set to true by default, and only change it to false if exception or count is higher than threshold.

Reply:
you are right, check next iteration